common/dictionary-db/dictionary-db.ts

```diff
@@ -0,0 +1,1439 @@
+        }
+        const allCardIds = notesInfo.flatMap((noteInfo) => noteInfo.cards);
+        const cardsModTime = (await anki.cardsModTime(allCardIds)).reduce((acc, cur) => {
+            acc.set(cur.cardId, cur.mod);
+            return acc;
+        }, new Map<number, number>()); // Reviews or suspension status are reflected in card mod time
+        if (cardsModTime.size !== allCardIds.length) {
+            throw new Error('Anki changed during cards record build, some cards mod time could not be retrieved.');
+        }
+
+        const existingAnkiNoteIdMap = await this._getAnkiCardsByNoteIdBulk(profile, noteIds);
+        const modifiedNotes: NoteInfo[] = [];
+        const modifiedCardIdsSet = new Set<number>();
+        for (const noteInfo of notesInfo) {
+            if (!Object.keys(noteInfo.fields).length || !noteInfo.cards.length) continue;
+            const noteId = noteInfo.noteId;
+            const modifiedAt = Math.max(
+                notesModTime.get(noteId)!,
+                ...noteInfo.cards.map((cardId) => cardsModTime.get(cardId)!)
+            );
+
+            let modified = false;
+            const existingAnkiCards = existingAnkiNoteIdMap.get(noteId);
+            for (const [track, ts] of trackStates.entries()) {
+                if (
+                    !Object.keys(noteInfo.fields).some(
+                        (f) =>
+                            ts.dt.dictionaryAnkiWordFields.includes(f) || ts.dt.dictionaryAnkiSentenceFields.includes(f)
+                    )
+                ) {
+                    continue;
+                }
+                const trackAnkiCards = existingAnkiCards?.filter((ankiCard) => ankiCard.track === track) ?? [];
+                if (
+                    trackAnkiCards.length !== noteInfo.cards.length ||
+                    !noteInfo.cards.every((cardId) => trackAnkiCards.some((ankiCard) => ankiCard.cardId === cardId)) ||
+                    !trackAnkiCards.every((ankiCard) => ankiCard.modifiedAt === modifiedAt)
+                ) {
+                    modified = true;
+                    break;
+                }
+            }
+            if (!modified) continue;
+
+            noteInfo.mod = modifiedAt;
+            modifiedNotes.push(noteInfo);
+            for (const cardId of noteInfo.cards) modifiedCardIdsSet.add(cardId);
+        }
+
+        const modifiedCardsDeck: Map<number, string> = new Map();
+        const modifiedCardIds = Array.from(modifiedCardIdsSet);
+        if (modifiedNotes.length) {
+            for (const cardInfo of await anki.cardsInfo(modifiedCardIds)) {
+                modifiedCardsDeck.set(cardInfo.cardId, cardInfo.deckName);
+            }
+            const dts = Array.from(trackStates.values()).map((ts) => ts.dt);
+            for (let i = modifiedNotes.length - 1; i >= 0; i--) {
+                if (
+                    !dts.some(
+                        (dt) =>
+                            (!dt.dictionaryAnkiDecks.length ||
```

Comment:
It might be because I'm getting old but logic like this takes me a while to read haha